### ISSUE TITLE
Serve from a different domain

### ISSUE THREAD

**Issue by ctron** [friendica/friendica#5895]
With Mastodon it is possible for serve a user from a different domain. e.g. my user is `me@domain.org` but I want to serve this from `me@social.domain.org`, but still reference the user as `me@domain.org`.

Also see: https://github.com/tootsuite/documentation/blob/master/Running-Mastodon/Serving_a_different_domain.md

Is there a way to achieve something similar with Friendica? I wasn't able to find anything in the docs or the FAQ.

**Comment by MrPetovan** [comment]
No, we don't have any support for this. Friendica has support for sub-folders, but not multiple domains. From the documentation itself, it sounds neat, very risky and not something that I would want  to receive regular support question about.

Additionally, this guide was written with OStatus in mind, not sure if/how it would work with ActivityPub. Maybe @annando can enlighten us about the actual feasibility.

**Comment by ctron** [comment]
> Friendica has support for sub-folders

So I could host my user `me@domain.org` from `domain.org/friendica`?

Because the documentation says:

> What won’t work very well is a mere subdirectory, e.g.: yourself.com/friendica.

**Comment by MrPetovan** [comment]
Friendica will work, and you will be able to interact with other Friendica nodes, but not with any other networks (GNU Social, Mastodon, etc...).

**Comment by ctron** [comment]
Too bad, I definitely want that :smiley:

**Comment by MrPetovan** [comment]
I didn't know about it, and it definitely is something I'd like given that my current webfinger address, `hypolite@friendica.mrpetovan.com` is rather unwieldy and I'd prefer `hypolite@mrpetovan.com` without having to move my Friendica server. So it may very well happen, I'm waiting on @annando's feedback.

**Comment by annando** [comment]
I would have to have a deeper look at it to ensure no ugly side effects.

**Comment by MrPetovan** [comment]
What is the soonest you can do that?

**Comment by annando** [comment]
I must confess that I want to finish some other stuff before even thinking about this.

**Comment by tobiasd** [comment]
How is it done in Mastodon? Some URL rewrite rules for the main domain to redirect the communication endpoints to the subdomain with the actual Mastodon installation?

**Comment by annando** [comment]
@MrPetovan yes.

**Comment by MrPetovan** [comment]
Great, thanks!

**Comment by tobiasd** [comment]
So then add a rewrite rule for the top domain for the .well-known path. For me this sounds like a _support request_ on how this rewrite rule should look like.

**Comment by annando** [comment]
I'm completely unsure about the impact. Currently we can resolve from the URL to the address format vice versa. Having different domains for this could have very ugly side effects.

**Comment by tobiasd** [comment]
Should not the communication partner, once to the connection is established automatically communicate with the correct location?

I want to start a communication with `bob@example.com`. So Friendica looks that user up in the `.well-known/host-meta` structure of `example.com`. `example.com` returns, thanks to the rewrite rule, that bob has an account at `social.example.com` and from that moment on Friendica knows the location of the communication endpoints are `social.example.com/something`.

**Comment by annando** [comment]
There is the way via the webfinger to resolve ```bob@example.com``` to ```social.example.com/something```, but no way to resolve ```social.example.com/something``` to ```bob@example.com```.

**Comment by MrPetovan** [comment]
Like @annando said, there are two part to this:
1. The incoming resolution of `bob@example.com` webfinger which is a matter of server configuration like you suggested.
2. The outgoing publication of `bob@example.com` in any `social.example.com` output. By default, Friendica wouldn't have any knowledge of 1., which means that the published webfinger would still be `bob@social.example.com`. We need an additional config key to tell Friendica to serve the profile URL with `social.example.com` but the webfinger with `example.com`.

The potential side-effect is having a unique account on `social.example.com` having two valid webfinger addresses, which may look like two different accounts with the same profile URL.

**Comment by tobiasd** [comment]
> There is the way via the webfinger to resolve ```bob@example.com```
> to ```social.example.com/something```, but no way to resolve
> ```social.example.com/something``` to ```bob@example.com```.

Which was not the question.

Can you write down the `.well-known/host-meta` file content required _only_ for resolving the 1st (bob@example.com -> bob@social.example.com) so that it can be tested and maybe find its way into the FAQ or other documentation.

**Comment by tobiasd** [comment]
> But this alone wouldn't be sufficient to ensure correct communication
> if someone used the `bob@example.com` webfinger address.

Why not? The platform used for the communication looks up the well-known informations and resolves the location of the inbox/push/drop/outbox/pull/fetch (how-ever you want to call these locations) and the delivers the messages to the correct locations.

**Comment by MrPetovan** [comment]
Because `social.example.com` would still output that the webfinger address for `bob` is `bob@social.example.com`, not `bob@example.com`. What information the remote server is supposed to prioritize?

**Comment by tobiasd** [comment]
The correct ¸bob@social.exa,ple.com` one. But still, for the ease of the users you can pass around the information, that you can just use ¸bob@example.com`.

**Comment by MrPetovan** [comment]
> The correct `bob@social.example.com` one

Of course, but between this assumption and the reality, there's a large gap and a variety of implementations. That's why Friendica should know about `example.com` even though it's installed on `social.example.com`. So that all the webfinger addresses are the same across all the outputs.

Simple example with the current Friendica setup:

https://friendica.mrpetovan.com/.well-known/webfinger?resource=acct:hypolite@mrpetovan.com gives

```json
{
	"subject": "acct:hypolite@friendica.mrpetovan.com",
	"aliases": ["https:\/\/friendica.mrpetovan.com\/~hypolite", "https:\/\/friendica.mrpetovan.com\/profile\/hypolite"],
	"links": [{
		"rel": "http:\/\/webfinger.net\/rel\/profile-page",
		"type": "text\/html",
		"href": "https:\/\/friendica.mrpetovan.com\/profile\/hypolite"
	}, {
		"rel": "self",
		"type": "application\/activity+json",
		"href": "https:\/\/friendica.mrpetovan.com\/profile\/hypolite"
	},
	...
	]
}
```
See how the `subject` property is different from the Webfinger address that was provided? How is the remote server supposed to guess that the response isn't authoritative from the corresponding input?

The `subject` should be `acct:hypolite@mrpetovan.com` as well, but Friendica has no way of guessing it should do the replacement unless there's a config key for it.

**Comment by MrPetovan** [comment]
In this case, yes, it should only use the result. But what if it doesn't? Webfinger addresses are supposed to be unique, and yet `.well-know/host-meta` will output the same information for `bob@social.example.com` and `bob@example.com`,which can yield unpredictable results depending on the remote implementation of Webfinger.

Beyond that, the config key would allow a Friendica admin to decide if both `.well-know/host-meta` calls should return either `bob@social.example.com` or `bob@example.com`.

**Comment by tobiasd** [comment]
> Webfinger addresses are supposed to be unique,

Then this is a wontfix as `bob@example.com` and `bob@social.example.com` are not unique entities but instead both are one account by Bob.

**Comment by MrPetovan** [comment]
Now I'm reading the spec and it doesn't say it has to be unique, hence the "alias" part. I'm guessing if we enable such shenanigans, the alias should contain both versions.

**Comment by shiribailem** [comment]
Chiming in here because it's a subset of a broader thing I was interested in: you should be able to know which webfinger is getting called from the host header?

Also, counter point to this design request: how about putting the regular site in a subfolder and having a redirect?

ie. example.com would redirect to example.com/site/, and the rewrite rules could specifically skip anything ^site/? To link a user to log in, they could link to example.com/login
